framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4631,6 +4613,442 @@ private void makeGroundTargetType(
+   * @return the captured type
+   */
+  public AnnotatedTypeMirror applyCaptureConversion(AnnotatedTypeMirror typeToCapture) {
+    TypeMirror capturedTypeMirror = types.capture(typeToCapture.getUnderlyingType());
+    return applyCaptureConversion(typeToCapture, capturedTypeMirror);
+  }
+
+  /**
+   * Apply capture conversion.
+   *
+   * <p>Capture conversion is the process of converting wildcards in a generic type to fresh type
+   * variables. See <a
+   * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-5.html#jls-5.1.10">JLS 5.1.10</a>
+   * for all the details.
+   *
+   * <p>If {@code type} is not a declared type or if it does not have any wildcard type arguments,
+   * this method returns {@code type}.
+   *
+   * @param type type to capture
+   * @param typeMirror type from the Java compiler that has captured wildcards, which are used as
+   *     the underlying type of the captured wildcards
+   * @return the captured type
+   */
+  public AnnotatedTypeMirror applyCaptureConversion(
+      AnnotatedTypeMirror type, TypeMirror typeMirror) {
+    if (!shouldCapture(type, typeMirror)) {
+      if (type.containsUninferredTypeArguments()
+          && typeMirror.getKind() == TypeKind.DECLARED
+          && type.getKind() == TypeKind.DECLARED) {
+        // If capture conversion is not applied because the type contains uninferred
+        // wildcards, then mark any uncaptured wildcards as "uninferred".
+        DeclaredType capturedTypeMirror = (DeclaredType) typeMirror;
+        AnnotatedDeclaredType uncapturedType = (AnnotatedDeclaredType) type;
+        for (int i = 0; i < capturedTypeMirror.getTypeArguments().size(); i++) {
+          TypeMirror capturedTypeArgTM = capturedTypeMirror.getTypeArguments().get(i);
+          AnnotatedTypeMirror uncapturedTypeArg = uncapturedType.getTypeArguments().get(i);
+          if (uncapturedTypeArg.getKind() != TypeKind.WILDCARD) {
+            continue;
+          }
+          if (TypesUtils.isCaptured(capturedTypeArgTM)
+              || capturedTypeArgTM.getKind() != TypeKind.WILDCARD) {
+            ((AnnotatedWildcardType) uncapturedTypeArg).setUninferredTypeArgument();
+          }
+        }
+      }
+      return type;
+    }
+
+    DeclaredType capturedTypeMirror = (DeclaredType) typeMirror;
+    AnnotatedDeclaredType uncapturedType = (AnnotatedDeclaredType) type;
+    final AnnotatedDeclaredType capturedType =
+        (AnnotatedDeclaredType) AnnotatedTypeMirror.createType(capturedTypeMirror, this, false);
+
+    captureTypeArgCopier.copy(uncapturedType, capturedType);
+
+    AnnotatedDeclaredType typeDeclaration =
+        (AnnotatedDeclaredType) getAnnotatedType(uncapturedType.getUnderlyingType().asElement());
+
+    // A mapping from type variable to its type argument in the captured type.
+    Map<TypeVariable, AnnotatedTypeMirror> typeVariableMapping = new HashMap<>();
+    // A mapping from the captured type to the annotated captured type.
```

Comment:
Regarding "the captured type":  I think this method `applyCaptureConversion` is capturing a single type, which I would call "the captured type".  Is this a mapping for the type variables of that type, rather than a mapping from whole captured types?  I may be confused, but please clarify.

Reply:
Below I've quoted two related comments that should be discussed together.
> The Javadoc of `TypesUtils.isCaptured` says "Returns whether a TypeVariable represents a captured type." but I think it is actually whether the type variable comes from a wildcard.  The type variable is part of a captured type but is not itself a captured type.  Although javac's `Type.java` poorly uses the name `isCaptured`, I think that TypesUtils should use something more descriptive, such as `isCapturedTypeVariable` or `isTypeVariableFromCaptureConversion` or the like.

> "each captured type": This method only captures one type. Are you talking about the type variables that were constructed to stand in for wildcards -- that is, any type argument that changed as a result of capture conversion? Maybe we need terminology for them, but "captured type" seems wrong. This naming choice may be related to my comment about `isCaptured`.

You've pointed out that I have been using "captured type" to mean both the type resulting from apply capture conversion and the "fresh" type variables created during capture conversion. Thanks, this is confusing and should be cleaned up.  I agree that it makes more sense that "captured type" means the type resulting from applying capture conversion to a type.  

I was calling the type variables created during capture conversion "captured types" because:

1. That's what comments in the Checker Framework called them.
2. `CapturedType` is the name of the javac class that implements these type variables.
3. Calling `toString()` on a `CapturedType` results in something like `capture#826 of ?`.
4. In javac error messages such as:
```
  where CAP#1 is a fresh type-variable:
    CAP#1 extends CharSequence from capture of ? extends CharSequence
```

I can fix 1. 2 is an implementation detail of javac that can be ignored, but 3 and 4 are what really makes it seem like these type variables should be called "captured types" or maybe "capture types".  So, I propose calling type variables created during capture conversion "captured type variables".  What do you think?  Is that still too confusing?

An alternative would be "fresh type-variable" which is used in javac error messages.

Reply:
I like your proposal of "captured type variable".  (I indirectly suggested in in the suggestion about `isCapturedTypeVariable`.)

(3 and 4 use "capture" but not "captured type", so I think they are as compatible with "captured type variable" as with "captured type".)